src/core/compile/check/mod.rs

```diff
@@ -1070,11 +1076,8 @@ impl<'a> Context<'a> {
             ast::StatementObj::TyAssign { target, ty, value } => {
                 match (as_assignment_target(target), value) {
                     (Some(target), Some(value)) => {
-                        let undeclared = self.get_undeclared(&target, loc)?;
-                        self.assign_order.push(Assign::Declare {
-                            undeclared,
-                            target: target.clone(),
-                        });
+                        let assign_i = self.assign_order.len();
+                        self.assign_order.push(Assign::Placeholder);
```

Comment:
How come we need to do this? If I'm understanding correctly `assign_i` is the index that `Placeholder` will be pushed at, and then later we do `self.assign_order[assign_i] = Assign::Declare {...}`. How come we can't just do `self.assign_order.push(Assign::Declare { ... }` instead of pushing the placeholder first?

Reply:
It's more of me being safe + a little bit lazy. The `Assign` can only be created after the inner expressions of the statement are checked, and I didn't want to think too hard to see whether more assignments could be generated by expressions. Pretty sure they can't be though, and if not, then you're right and this can be refactored.